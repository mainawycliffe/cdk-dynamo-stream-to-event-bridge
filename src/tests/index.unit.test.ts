import {
  haveResourceLike,
  expect as expectCDK,
  SynthUtils,
} from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { Table, AttributeType, StreamViewType } from "@aws-cdk/aws-dynamodb";
import DynamoStreamToEventBridge from "..";

class TestStack extends Stack {
  constructor() {
    super();
    const table = new Table(this, "testStack", {
      tableName: "TableName",
      partitionKey: {
        name: "id",
        type: AttributeType.STRING,
      },
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
    });

    new DynamoStreamToEventBridge(this, "streamTable", {
      table,
      eventPrefix: "EventPrefix",
    });
  }
}

const testStack = new TestStack();

it("creates a Lambda function to process the Dynamo stream and emit EventBridge events", () => {
  expectCDK(testStack).to(
    haveResourceLike("AWS::Lambda::Function", {
      Environment: {
        Variables: {
          AWS_NODEJS_CONNECTION_REUSE_ENABLED: "1",
          EVENT_BUS_ARN: {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  Ref: "AWS::Partition",
                },
                ":events:",
                {
                  Ref: "AWS::Region",
                },
                ":",
                {
                  Ref: "AWS::AccountId",
                },
                ":event-bus/default",
              ],
            ],
          },
          EVENT_PREFIX: "EventPrefix",
        },
      },
      Handler: "index.handler",
      MemorySize: 1024,
    })
  );
});

it("creates an event source mapping", () => {
  expectCDK(testStack).to(haveResourceLike("AWS::Lambda::EventSourceMapping"));
});

it("creates an Alarm to alert on errors", () => {
  expectCDK(testStack).to(haveResourceLike("AWS::CloudWatch::Alarm"));
});

it("creates resources matching the snapshot", () => {
  expect(SynthUtils.toCloudFormation(testStack)).toMatchSnapshot();
});
