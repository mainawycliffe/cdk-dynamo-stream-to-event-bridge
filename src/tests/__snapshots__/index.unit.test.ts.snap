// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "streamTableAlarm30FED3DA": Object {
      "Properties": Object {
        "AlarmDescription": "Alert when a record has failed to stream to event bridge using cdk-dynamo-stream-to-event-bridge.",
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "testStack53BF581B",
              },
              "StreamToEventBridge",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "Dimensions": Array [
          Object {
            "Name": "QueueName",
            "Value": Object {
              "Fn::GetAtt": Array [
                "streamTableDLQ0AE2E98D",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "NumberOfMessagesSent",
        "Namespace": "AWS/SQS",
        "Period": 3600,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "streamTableDLQ0AE2E98D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "MessageRetentionPeriod": 1209600,
        "QueueName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "testStack53BF581B",
              },
              "StreamToEventBridgeDLQ",
            ],
          ],
        },
        "VisibilityTimeout": 120,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "streamTableFunctionDynamoDBEventSourcetestStack45E4763F": Object {
      "Properties": Object {
        "BatchSize": 10,
        "BisectBatchOnFunctionError": true,
        "DestinationConfig": Object {
          "OnFailure": Object {
            "Destination": Object {
              "Fn::GetAtt": Array [
                "streamTableDLQ0AE2E98D",
                "Arn",
              ],
            },
          },
        },
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "testStack53BF581B",
            "StreamArn",
          ],
        },
        "FunctionName": Object {
          "Ref": "streamTableFunctionE2FC641D",
        },
        "MaximumRetryAttempts": 10,
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "streamTableFunctionE2FC641D": Object {
      "DependsOn": Array [
        "streamTableFunctionServiceRoleDefaultPolicyA0B5635E",
        "streamTableFunctionServiceRole077F8269",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-lambda-nodejs-mock-fake-bucket",
          "S3Key": "cdk-lambda-nodejs-mock-fake-lambda.ts",
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Streams updated rows from the table ",
              Object {
                "Ref": "testStack53BF581B",
              },
              "to EventBridge.",
            ],
          ],
        },
        "Environment": Object {
          "Variables": Object {
            "EVENT_BUS_ARN": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":events:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":event-bus/default",
                ],
              ],
            },
            "EVENT_PREFIX": "EventPrefix",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "streamTableFunctionServiceRole077F8269",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "streamTableFunctionServiceRole077F8269": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "streamTableFunctionServiceRoleDefaultPolicyA0B5635E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":events:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":event-bus/default",
                  ],
                ],
              },
            },
            Object {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testStack53BF581B",
                  "StreamArn",
                ],
              },
            },
            Object {
              "Action": Array [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "streamTableDLQ0AE2E98D",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "streamTableDLQ0AE2E98D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "streamTableFunctionServiceRoleDefaultPolicyA0B5635E",
        "Roles": Array [
          Object {
            "Ref": "streamTableFunctionServiceRole077F8269",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "streamTableFunctionstreamTableDLQMapping07DEC73C": Object {
      "Properties": Object {
        "BatchSize": 10,
        "Enabled": false,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "streamTableDLQ0AE2E98D",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "streamTableFunctionE2FC641D",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "testStack53BF581B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "TableName",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
